File: OstiumFrontend/OstiumLogic/OstiumBackend/GameSetup/Impl/Logic.cs
using System;
using System.Collections.Generic;
using System.Linq;
using GameComponents.Api;
using GameSetup.Api;

namespace GameSetup.Impl
{
    public class GameSetupApiLogic : GameSetupApi
    {
        private List<CreatureCard> hand = new List<CreatureCard>
        {
            CreatureCard.Create(new CreatureCardId("Mouse1")),
            CreatureCard.Create(new CreatureCardId("Mouse2"))
        };

        private CreatureCard? attackRow = null;
        private CreatureCard? defenseRow = null;

        public Game StartGame()
        {
            return ToApiGame();
        }

        public Game PlayCard(CreatureCardId cardId, RowType row)
        {
            var card = hand.First(c => c.GetId().Equals(cardId));
            hand.Remove(card);

            switch (row)
            {
                case RowType.ATTACK:
                    attackRow = CreatureCard.Create(cardId);
                    break;
                case RowType.DEFENSE:
                    defenseRow = CreatureCard.Create(cardId);
                    break;
            }

            return ToApiGame();
        }

        public Game MoveCard(CreatureCardId cardId, RowType from, RowType to)
        {
            CreatureCard? card = from switch
            {
                RowType.ATTACK when attackRow?.GetId().Value == cardId.Value => attackRow,
                RowType.DEFENSE when defenseRow?.GetId().Value == cardId.Value => defenseRow,
                _ => throw new InvalidOperationException($"No card with ID {cardId} in {from} row")
            };

            // Remove the card from the 'from' row
            switch (from)
            {
                case RowType.ATTACK:
                    attackRow = null;
                    break;
                case RowType.DEFENSE:
                    defenseRow = null;
                    break;
            }

            // Place the card in the 'to' row
            switch (to)
            {
                case RowType.ATTACK:
                    defenseRow = attackRow;
                    attackRow = card;
                    break;
                case RowType.DEFENSE:
                    attackRow = defenseRow;
                    defenseRow = card;
                    break;
            }

            return ToApiGame();
        }


        private Game ToApiGame()
        {
            return Game.Create(
                new Table(
                    GateDurabilityCard.Create(
                        new GateDurabilityMarker(15),
                        new GateDurabilityMarker(15)
                    ),
                    attackRow,
                    defenseRow,
                    GateCard.Create(false)
                ),
                Hand.Create(hand)
            );
        }
    }
}

File: Backend/lib/src/main/kotlin/com/github/bratek20/ostium/gamesetup/impl/Logic.kt
package com.github.bratek20.ostium.gamesetup.impl

import com.github.bratek20.logs.api.Logger
import com.github.bratek20.ostium.gamesetup.api.*

import com.github.bratek20.ostium.gamecomponents.api.*

class GameSetupApiLogic(
    private val logger: Logger
): GameSetupApi {
    private val hand: MutableList<CreatureCard> = mutableListOf(
        CreatureCard.create(
            id = CreatureCardId("Mouse1"),
        ),
        CreatureCard.create(
            id = CreatureCardId("Mouse2"),
        ),
    )
    private var attackRow: CreatureCard? = null
    private var defenseRow: CreatureCard? = null

    override fun startGame(): Game {
        logger.info("Game started")
        return toApiGame()
    }

    override fun playCard(cardId: CreatureCardId, row: RowType): Game {
        val card = hand.first { it.getId() == cardId }
        hand.remove(card)

        when (row) {
            RowType.ATTACK -> attackRow = CreatureCard.create(id = cardId)
            RowType.DEFENSE -> defenseRow = CreatureCard.create(id = cardId)
        }

        return toApiGame()
    }

    override fun moveCard(cardId: CreatureCardId, from: RowType, to: RowType): Game {
        val card = when (from) {
            RowType.ATTACK -> attackRow
            RowType.DEFENSE -> defenseRow
        } ?: throw IllegalStateException("No card in $from row")

        when (to) {
            RowType.ATTACK -> attackRow = card
            RowType.DEFENSE -> defenseRow = card
        }

        when (from) {
            RowType.ATTACK -> attackRow = null
            RowType.DEFENSE -> defenseRow = null
        }

        return toApiGame()
    }

    private fun toApiGame(): Game {
        return Game.create(
            hand = Hand.create(
                cards = hand.toList()
            ),
            table = Table.create(
                gateDurabilityCard = GateDurabilityCard.create(
                    myMarker = GateDurabilityMarker(15),
                    opponentMarker = GateDurabilityMarker(15)
                ),
                attackRow = attackRow,
                defenseRow = defenseRow,
                gateCard = GateCard.create(false),
            )
        )
    }
}